Update CI to include Ruby 3.1

petergoldstein opened the pull request:
This PR updates the CI configuration to include Ruby 3.1

Other changes include:

1. Updating the wetransfer_style gem to a ~> 1.0 version
2. Removing the 2.1 in the CI configuration, as that is not supported by the underlying Rubocop version
3. Addressing all open lints from the wetransfer_style upgrade
4. Adding JRuby 9.3
5. Making a change in a spec to avoid use of a constructor for Time that doesn't exist in Ruby 3.1

grdw commented on .github/workflows/ci.yml:
```diff
@@ -56,15 +56,15 @@ jobs:
+          - '3.1'
+          - '3.0'
           - '2.7'
-          - '2.1'
```
There are possibly still other Ruby engineers who depend on Ruby 2.1. Perhaps leave this out for now, and we'll create a ticket for deprecating support for Ruby 2.1.

grdw commented on .rubocop.yml:
```diff
@@ -1,7 +1,7 @@
 inherit_gem:
   wetransfer_style: ruby/default.yml
 AllCops:
-  TargetRubyVersion: 2.1
```
Same thing as mentioned above.

petergoldstein replied:
Can't do that and make the PR work.  There's no version of Rubocop that runs on Ruby 3.1 that supports Ruby 2.1 styling.

Keep in mind that Ruby 2.1 EOLd on March 31, 2017 - almost 5 years ago.

grdw commented on examples/rack_application.rb:
```diff
@@ -36,20 +36,19 @@ def call(env)
-        # here directly to the Rack server handler.
-        IO.copy_stream(f, zip)
+    zip_response_body =
+      # We are adding only one file to the ZIP here, but you could do that
```
Indentation wise, this could be cleaned up I feel. F.e. the contents of the block is now 2 tabs (or 4 spaces) behind the start of the block.

Favour doing this:
```ruby
ZipTricks::Streamer.output_enum do |zip|
  zip.add_stored_entry(..)
end
```

over:

```ruby
    ZipTricks::Streamer.output_enum do |zip|
  zip.add_stored_entry(..)
end
```

grdw commented on lib/zip_tricks/size_estimator.rb:
```diff
@@ -43,9 +43,7 @@ def add_stored_entry(filename:, size:, use_data_descriptor: false)
-    if use_data_descriptor
-      @streamer.update_last_entry_and_write_data_descriptor(crc32: 0, compressed_size: size, uncompressed_size: size)
-    end
+    @streamer.update_last_entry_and_write_data_descriptor(crc32: 0, compressed_size: size, uncompressed_size: size) if use_data_descriptor
```
Style-wise, it's preferred to use the previous code as it was, because of the length of the code line. If Rubocop is raising a problem here, feel free to add an ignore statement.

grdw commented on lib/zip_tricks/stream_crc32.rb:
```diff
@@ -5,7 +5,7 @@ class ZipTricks::StreamCRC32
     String.new('', capacity: 1)
     true
-  rescue ArgumentError
+                                  rescue ArgumentError
```
Suggestion, I would change this to:

```ruby
STRINGS_HAVE_CAPACITY_SUPPORT = 
  begin
    # ...
  rescue ArgumentError
    # ...
  end
```

grdw commented on lib/zip_tricks/zip_writer.rb:
```diff
@@ -104,9 +104,7 @@ def write_local_file_header(io:, filename:, compressed_size:, uncompressed_size:
-    if requires_zip64
-      extra_fields << zip_64_extra_for_local_file_header(compressed_size: compressed_size, uncompressed_size: uncompressed_size)
-    end
+    extra_fields << zip_64_extra_for_local_file_header(compressed_size: compressed_size, uncompressed_size: uncompressed_size) if requires_zip64
```
Same thing as described earlier.

grdw commented on spec/support/allocate_under_matcher.rb:
```diff
@@ -2,9 +2,7 @@
-    unless RUBY_ENGINE == 'ruby'
-      skip "allocation tracing not supported on #{RUBY_ENGINE}"
-    end
+    skip "allocation tracing not supported on #{RUBY_ENGINE}" unless RUBY_ENGINE == 'ruby'
```
Here, it's okay to keep the `unless` on a single line, after the statement, considering the line-length is shorter.

grdw commented on spec/support/managed_tempfile.rb:
```diff
@@ -8,12 +8,10 @@ def initialize(*)
     @@managed_tempfiles.each do |tf|
-      begin
         tf.close
         tf.unlink
```
Fix indentation. There's now 4 spaces between the start of the block and the `rescue`.

grdw commented on spec/zip_tricks/remote_uncap_spec.rb:
```diff
@@ -32,10 +32,8 @@
 
   after :each do
-    begin
       File.unlink('temp.zip')
```
Same issue as above.

grdw commented:
Thanks for your contributions. I added a few comments, but most of these are styling problems I feel that come from Rubocop.

courtsimas commented:
👀

petergoldstein commented:
To get this to run green:

1. Updated wetransfer_style to ~>2.1
2. Loosened version constraints for rake, rack
3. Updated the ubuntu-18.04 image to ubuntu-20.04
4. Removed JRuby 9.2 from CI
5. Addressed a few new lints from the wetransfer_style upgrade
6. Updated to checkout/actions@v4

Runs green on my fork.
